### PR TITLE
Add Digital Ocean Droplet module

### PR DESCRIPTION
**Pull request by gurch101**
##### SUMMARY
Create a new digital ocean droplet API to remove dopy dependency.
Also enable all digital ocean v2 parameters for droplet creation.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
digital_ocean_droplet

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/gurch/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

### CODE REVIEW

**Comment by Akasurde** [comment]
@gurch101 Thanks for your PR. Would you please re-consider using existing digital_ocean module for refactoring ? Creating a new module may lead to confusion in user.

Let me know your opinion. cc @BondAnthony

**Comment by BondAnthony** [comment]
@Akasurde I believe this would be the right approach. The community of Digital Ocean Ansible users is well aware that `digital_ocean` module requires the use of `dopy`. In addition to managing droplets the `digital_ocean` module also managed ssh keys on the `v1` api. There has been a lot of talk within Ansible pull requests about creating a `digital_ocean_droplet` module that performs droplet only actions and moving ssh, tagging and other actions into their own modules.

I think this would be the right direction along with deprecating `digital_ocean` once merged. Old issue about deprecating `digital_ocean` #29740

@gurch101 thank you for contribution! Could you please address the comments from the bot.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: Please include the `DO_OAUTH_TOKEN` environment variable as one of the `env_fallback` values.

**Comment by Akasurde** [comment]
@BondAnthony Thanks for your views. I am OK with approach of deprecating `digital_ocean` and adding `digital_ocean_droplet`.

Please also add details about this deprecation [here](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/porting_guide_2.5.rst#modules-removed).

**Comment by gundalow** [comment]
CI still failing
```
Command "test/sanity/code-smell/boilerplate.sh" returned exit status 2.
>>> Standard Output
== Missing __metaclass__ = type ==
./lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py

== Missing from __future__ import (absolute_import, division, print_function) ==
./lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
```


```
lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:0:0: E305 DOCUMENTATION.version_added: required key not provided @ data['version_added']. Got None
lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:0:0: E307 version_added should be 2.5. Currently 0.0
lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:0:0: E312 No RETURN provided
lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:0:0: E316 ANSIBLE_METADATA.metadata_version: not a valid value for dictionary value @ data['metadata_version']. Got '1.0'
```

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: I feel like 300 seconds is a long time to wait. Most droplets are created within 120 second range.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: I believe `backups_enabled`is duplicated in the doc string. Should this be removed since you listed `backups` on line 88 and use the backups parameter in the argument spec?

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:108]: Please include the `monitoring` option as well. This will allow users to create droplets with [enhanced monitoring](https://developers.digitalocean.com/documentation/v2/#create-a-new-droplet) from the beginning instead of forcing them to add the solution later on.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: `Create and delete a DigitalOcean droplet`

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: `DigitalOcean OAuth token. Can be specified in C(DO_API_KEY) or C(DO_API_TOKEN) environment variables`

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: Examples should have `names:` rather than comments

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: name

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: `default=False`
(same for others)

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:314]: Please add droplet_id to `aliases:` in `DOCUMENTATION`

http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: Since ssh key management is no longer being performed within this module please remove the ssh key reference.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: Please add `monitoring` and pass this argument to the request.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: You will need to implement links when performing this GET operation on droplets. Please take a look at the [links](https://developers.digitalocean.com/documentation/v2/#links) documentation.

**Comment by Akasurde** [comment]
@gurch101 Are you still working on this ? 

needs_info

**Comment by Akasurde** [comment]
@gurch101 Are you still working on this? If not, let me know I will take over this PR.

**Comment by gurch101** [comment]
@Akasurde apologies! I will try to get this done in next few days

**Comment by varunchopra** [comment]
I think it's a good idea to use lists for names, and call `/v2/droplets` instead.

How do you want to deal with regions? This will just throw errors if a region is unavailable.

**Comment by gurch101** [comment]
@chopraaa 
what use case do you have in mind for using a list of names? Would using a loop in a playbook not suffice?

This code uses the DigitalOceanHelper module which ultimately uses the /v2/.

Haven't tried passing in an invalid region but isn't throwing an error precisely what we would want to do?

**Comment by Gunni** [comment]
Please add IPv6 addresses to the result aswell

    result.droplet.ip_address: <ipv4 address to be backwards compatible>
    result.droplet.ip_addresses: <list of ipv6 and ipv4 addresses>
    result.droplet.ipv4_address: <ipv4 address>
    result.droplet.ipv6_address: <ipv6 address>

Unset if ipv6 is not requested.

Additionally, why not enable ipv6 by default?

**Comment by Gunni** [comment]
Please add vm tags to the droplet create syntax.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: Please update your example with a valid task. `Command` is not a valid argument to the module anymore.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:256]: @gurch101 when a region is not available the following response is sent back with a status code of `422`. I think you will need to catch those errors and fail.
```
{
  'message': 'Region is not available', 
  'id': 'unprocessable_entity'
}
```

Digital Ocean currently doesn't supply a failover region, so this should result in a `self.module.fail_json(changed=False, msg='Region currently not available')`

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: Please update the version add to `2.8`

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:247]: @gurch101 here is suggested change to the create method based on the feedback from other individuals. We do need to include the ip addresses at the top level of the response so they can be easily passed into additional tasks or facts. Let me know what you think of this method change..

```python
    def create(self):
        json_data = self.get_droplet()
        droplet_data = None
        if json_data:
            droplet_data = self.get_addresses(json_data)
            self.module.exit_json(changed=False, data=droplet_data)
        response = self.rest.post('droplets', data=self.module.params)
        json_data = response.json
        if response.status_code == 422 and json_data['message'] == 'Region is not available':
            self.module.fail_json(chnaged=False, msg=json_data['message'])
        if self.wait:
            json_data = self.ensure_power_on(json_data['droplet']['id'])
            droplet_data = self.get_addresses(json_data)
        self.module.exit_json(changed=True, data=droplet_data)
```

The address method is used to pull out the different ip addresses DO sends back in the response.

```python
    def get_addresses(self, data):
        """
         Expose IP addresses as their own property allowing users extend to additional tasks
        """
        _data = data
        for k, v in data.items():
            setattr(self, k, v)
        networks = _data['droplet']['networks']
        for network in networks.get('v4', []):
            if network['type'] == 'public':
                _data['ip_address'] = network['ip_address']
            else:
                _data['private_ipv4_address'] = network['ip_address']
        for network in networks.get('v6', []):
            if network['type'] == 'public':
                _data['ipv6_address'] = network['ip_address']
            else:
                _data['private_ipv6_address'] = network['ip_address']
        return _data
```

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:147]: Please also update the response document to represent the ip addresses:

```
        "ip_address": "104.248.118.172",
        "ipv6_address": "2604:a880:400:d1::90a:6001",
        "private_ipv4_address": "10.136.122.141",
        "droplet": {....}
```

**Review by BondAnthony — COMMENTED** [review]
@gurch101 Please let me know if you have any questions. I've uploaded my changes and tested the module a decent amount. Great job on refactoring this into a droplet specific module!

Full patch:
https://github.com/BondAnthony/ansible/commit/762b8abe2e56d6efe2a9e27f6902382df0d4852b

**Comment by BondAnthony** [comment]
@Gunni I'm not sure I understand the point of including the `ip_addresses` list with a mix of both ipv4 and ipv6 addresses when the addresses are already being listed in the droplet.network object. Currently the DO platform will always return the ipv4 address as it's standard public ip. Since ipv6 is an added feature and by default disabled, I think it would be best to not enable ipv6 as a default.

**Comment by gurch101** [comment]
@BondAnthony thanks so much for the updates. Not sure what the convention is for this project - would you like me to close this PR, add your commit on top of mine, or should I just squash your patch into my commit?

**Comment by BondAnthony** [comment]
@gurch101 My changes were just a suggestion by no means do you need to include them. I’m fine with any direction you would like to take. I personally would include the changes and squash the commits.

**Comment by pokonski** [comment]
This would be great to have as it's not possible to use tagging functionality with the stable version of this module

**Comment by BondAnthony** [comment]
@gurch101 could you fix the example by removing command parameter?

**Comment by BondAnthony** [comment]
ping @gurch101

**Comment by gurch101** [comment]
@BondAnthony should be good now. Thanks!

**Comment by BondAnthony** [comment]
Thank you @gurch101 

shipit

**Comment by BondAnthony** [comment]
bot_status

**Comment by BondAnthony** [comment]
@Akasurde when you have time could you please run rebuild_merge. Do we know why the bot keeps getting stuck?

**Comment by pilou-** [comment]
Hello! @Akasurde, @pmarques (as randomly selected maintainer of `cloud/digital_ocean` namespace) could you please review this contribution?

**Review by pilou- — COMMENTED** [review]
Could you add support to `check_mode` ? This requires:
* add  `supports_check_mode=True` to `AnsibleModule`
* in both `create`/`delete` methods, once `json_data` is defined, test for `check_mode` and call `exit_json` when check mode is enabled ([example](https://github.com/ansible/ansible/blob/dcb35c427092b3bc2ef8f9b9e1132a3184b9e20b/lib/ansible/modules/cloud/digital_ocean/digital_ocean_sshkey.py#L181))

**Comment by gurch101** [comment]
@pilou- `check_mode` was added. Thanks.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (bcoca) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: catchall exceptions make for bad UI, we normally prefer a user friendly message, detailing what failed and hinting possible resolutions vs passing on API messages that are normally only useful to programmers

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by BondAnthony** [comment]
Hey @gurch101 when you have time could you address the comments, please.

**Comment by kbreit** [comment]
Are there plans to include integration tests with this?

**Comment by pmarques** [comment]
@kbreit There was some discussion here https://github.com/ansible/ansible/pull/29960

**Comment by gundalow** [comment]
Hi, Seems like there are a few related PRs floating around, great work :)

What's the plan, is it:
1) Merge this, as-is
2) Merge https://github.com/ansible/ansible/pull/47272 deprecation, as-is
3) https://github.com/ansible/ansible/pull/29960 Tests review and merge

Anything left before we merge this?

**Comment by BondAnthony** [comment]
@gundalow I would like for this PR to be merged after core approves and then we should be good to merge the #47272.

#29960 is waiting for Matt to approval unless you can do that.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [reply lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:108]: Looks like `monitoring:` has been added.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [reply lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:147]: Looks like IP address information has been added

**Comment by gundalow** [comment]
@BondAnthony Thanks for the info. Could you mark and completed items as `resolved`, that should make it easier to confirm what's left to fix.

Looks like there are a few updates needed to be made by @gurch101

**Comment by gurch101** [comment]
I personally feel that displaying errors straight from digital ocean is a reasonable thing to do since the few that I encountered while building this were readable. I'm a bit busy with work but I'll run through this again over the weekend. Seems like checking if the status code != 201/204 might lead to more granular error handling rather than having a top-level catch all try/catch.

thanks everyone for the feedback

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gurch101) [reply lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: I added https://github.com/ansible/ansible/blob/83df7088e1d13e84caec7f06495b1324ea1a7cd7/lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py#L258

to handle any errors on create. This catches invalid image names, regions, and sizes.

I realize this probably isn't what you wanted based on your comment about but in all of these cases, DigitalOcean returns an HTTP 422 + a message. Short of parsing the message itself or displaying a generic error message, I just relay it back to the user. IMO, the messages that come from DigitalOcean describe what failed in a user-friendly way.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (phemmer) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: This line is causing the module to fail unconditionally on every create.

**Review by phemmer — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
@gurch101 Hi, when you get a chance would you please be able to address the final few comments? Thanks in advance!

Once this is merged we can get https://github.com/ansible/ansible/pull/47272 in which should help make future development easier.

**Comment by gundalow** [comment]
@BondAnthony Would you please be able to re-review? Could you mark and completed items as resolved, that should make it easier to confirm what's left to fix.

**Review by bcoca — APPROVED** [review]

**Review by BondAnthony — APPROVED** [review]

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (BondAnthony) [reply lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py]: Looks like we just need this line removed and we should be good to merge.

**Comment by BondAnthony** [comment]
@gundalow everything looks good minus the removal of `line:257`. We should be good to merge this!

**Comment by kbreit** [comment]
Should a deprecation note be added to `digital_ocean` or should that be another PR?

**Comment by BondAnthony** [comment]
@kbreit PR #47272 has this information.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py
Comment (gundalow) [reply lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:256]: @BondAnthony Is this resolved?

**Comment by gundalow** [comment]
Thank you everybody, this is now merged!

I've also merged `DigitalOcean module deprecation` #47272 

Now we can focus on the rest of https://github.com/ansible/ansible/labels/digital_ocean

**Comment by nerijus** [comment]
When trying example (I added correct oauth_token etc):
```
- name: create a new droplet
  digital_ocean_droplet:
    state: present
    name: mydroplet
    oauth_token: XXX
    size: 2gb
    region: sfo1
    image: ubuntu-16-04-x64
    wait_timeout: 500
  register: my_droplet

- debug:
    msg: "ID is {{ my_droplet.droplet.id }}"
```
I get:
```
fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'droplet'
```
when trying to get my_droplet.droplet.id.

**Comment by nerijus** [comment]
The following line works:
```
msg: "ID is {{ my_droplet.data.droplet.id }} IP is {{ my_droplet.data.ip_address }} IPv6 is {{ my_droplet.data.ipv6_address }}"
```
I'd suggest updating example.

**Pull request merged**
